components/datetime/src/provider/packed_pattern.rs

```diff
@@ -0,0 +1,480 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! Data structures for packing of datetime patterns.
+
+use crate::{
+    helpers::size_test,
+    pattern::{
+        runtime::{PatternBorrowed, PatternMetadata},
+        PatternItem,
+    },
+    NeoSkeletonLength,
+};
+use alloc::vec::Vec;
+use icu_plurals::{provider::PluralElementsPackedULE, PluralElements};
+use zerovec::{VarZeroVec, ZeroSlice};
+
+use crate::pattern::runtime::Pattern;
+
+/// A field of [`PackedPatternsBuilder`].
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct LengthPluralElements<T> {
+    /// The "long" length pattern plural elements.
+    pub long: PluralElements<T>,
+    /// The "medium" length pattern plural elements.
+    pub medium: PluralElements<T>,
+    /// The "short" length pattern plural elements.
+    pub short: PluralElements<T>,
+}
+
+/// A builder for a [`PackedPatternsV1`].
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct PackedPatternsBuilder<'a> {
+    /// Patterns always available.
+    pub standard: LengthPluralElements<Pattern<'a>>,
+    /// Patterns for variant 0. If `None`, falls back to standard.
+    pub variant0: Option<LengthPluralElements<Pattern<'a>>>,
+    /// Patterns for variant 1. If `None`, falls back to standard.
+    pub variant1: Option<LengthPluralElements<Pattern<'a>>>,
+}
+
+size_test!(PackedPatternsV1, packed_skeleton_data_size, 32);
+
+/// Main data struct for packed datetime patterns.
+///
+/// ## Variants
+///
+/// This supports a set of "standard" patterns plus up to two "variants".
+/// The variants are currently used by year formatting:
+///
+/// - Standard: Year, which could be partial precision (2-digit Gregorain)
+/// - Variant 0: Full Year, which is always full precision
```

Comment:
Is your suggestion to call them `full_year` and `with_era`? Because I wanted these to be generalizable enough to other cases that might want to use this mechanism in the future. For instance, maybe some locale will want to use it for day periods or something ("6 in the evening", "7 in the evening", "8pm", "9pm").